kubernetes/config/kube_config.py

```diff
@@ -22,151 +22,243 @@
+
+    def _load_authentication(self):
+        """Read authentication from kube-config user section if exists.
+
+        This function goes through various authentication methods in user
+        section of kube-config and stops if it finds a valid authentication
+        method. The order of authentication methods is:
+
+            1. GCP auth-provider
+            2. token_data
+            3. token field (point to a token file)
+            4. username/password
+        """
+        if not self._user:
+            return
+        if self._load_gcp_token():
+            return
+        if self._load_user_token():
+            return
+        self._load_user_pass_token()
+
+    def _load_gcp_token(self):
+        if 'auth-provider' not in self._user:
+            return
+        if 'name' not in self._user['auth-provider']:
+            return
+        if self._user['auth-provider']['name'] != 'gcp':
+            return
+        # Ignore configs in auth-provider and rely on GoogleCredentials
+        # caching and refresh mechanism.
+        # TODO: support gcp command based token ("cmd-path" config).
+        self.token = self._get_google_credentials()
+        return self.token
+
+    def _load_user_token(self):
+        token = FileOrData(self._user, 'tokenFile', 'token').as_data()
+        if token:
+            self.token = token
+            return True
+
+    def _load_user_pass_token(self):
+        if 'username' in self._user and 'password' in self._user:
+            self.token = urllib3.util.make_headers(
+                basic_auth=(self._user['username'] + ':' +
+                            self._user['password'])).get('authorization')
+            return True
+
+    def _load_cluster_info(self):
+        if 'server' in self._cluster:
+            self.host = self._cluster['server']
+            if self.host.startswith("https"):
+                self.ssl_ca_cert = FileOrData(
+                    self._cluster, 'certificate-authority').as_file()
+                self.cert_file = FileOrData(
+                    self._user, 'client-certificate').as_file()
+                self.key_file = FileOrData(self._user, 'client-key').as_file()
+
+    def _set_config(self):
+        if 'token' in self.__dict__:
+            self._client_configuration.api_key['authorization'] = self.token
+        # copy these keys directly from self to configuration object
+        keys = ['host', 'ssl_ca_cert', 'cert_file', 'key_file']
+        for key in keys:
+            if key in self.__dict__:
+                setattr(self._client_configuration, key, getattr(self, key))
+
+    def load_and_set(self):
+        self._load_authentication()
+        self._load_cluster_info()
+        self._set_config()
+
+    def list_contexts(self):
+        return [context.value for context in self._config['contexts']]
+
+    @property
+    def current_context(self):
+        return self._current_context.value
+
+
+class ConfigNode:
```

Comment:
If this is going to be used in python 2.7, can this be a new style class please?